docs/dev/guidelines/server.rst

```diff
@@ -0,0 +1,148 @@
+
+Additional notes on the controller methods:
+
+* POST should return the newly created entity
+* Always use @PreAuthorize tag to only allow certain roles to access the method.
+* Never trust user input and check if the passed data exists in the database.
+* Always use different response status codes to notify the client about errors on the server:
+    * Forbidden - the user is not authorized to access the controller.
+    * Bad Request - the request was wrong.
+    * Not Found - can't find the requested data or it should be not accessible yet.
+
+12. Dependency injection:
+=========================
+
+* Some of you may argue this, but by favoring constructor injection you can keep your business logic free from Spring. Not only is the @Autowired annotation optional on constructors, you also get the benefit of being able to easily instantiate your bean without Spring.
+* Use setter based DI only for optional dependencies.
+* Avoid circular dependencies, try constructor and setter based DI for such cases.
+
+13. REST best practices:
+========================
+
+* Verify that API endpoints perform appropriate authorization and authentication consistent with the rest of the code base.
+* Check for other common weaknesses, e.g., weak configuration, malicious user input, missing log events, etc.
+* Handle exceptions and errors with a standard response. Errors are very important in REST APIs. They inform clients that something went wrong, after all.
+
+14. Keep it simple and stupid:
+==============================
+
+* Don’t write complex code.
+* Don’t write code when you are tired or in a bad mood.
+* Optimization vs Readability, always write code that is simple to read and which will be understandable for developers. Because time and resources that will be spent on hard readable code will be much higher than what you get from optimization.
+* Commit messages should describe both what the commit changes and how it does it.
+* ARCHITECTURE FIRST, writing code without thinking of its architecture is useless in the same way as dreaming about your desires without a plan of achieving them.
+
+15. General best practices:
+===========================
+
+* Always use the least possible access level, prefer using private over public access modifier (protected can be used as well).
+* Previously we used transactions very random, now we have to avoid using Transactional, if not sure that they are needed. Transactions can kill performance, introduce locking issues and database concurrency problems, and add complexity to our application. Good read: https://codete.com/blog/5-common-spring-transactional-pitfalls/
+* Define a constant if the same value is used more than once. Constants allow to change code later a lot easier, instead of looking for the places where this variable was used, you only need to change it in only one place.
```

Comment:
```suggestion
* Define a constant if the same value is used more than once. Constants allow you to change code later a lot easier. Instead of looking for the places where this variable was used, you only need to change it in only one place.
```